### PR TITLE
Issue#249_Refactoring_Quartz_job_creator

### PR DESCRIPTION
**Pull request by Illia-Shchur**
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009)

## Code reviewers

- [ ] @kh0ma
- [ ] @likeRewca
- [ ] @Ilya-Ross
- [x] @AriannaLiss
- [ ] @mitianin
- [ ] @IhorSamoshost
- [ ] @kizimov
- [x] @Vladyslav-Frolov

## Summary of issue

Needed easier way to create jobs and configure their cron at runtime
 using spring actuator.

## Summary of change

Added annotation QuartzJobBeanAutoConfiguration with BeanFactoryPostProcessor which configures job creating trigger and jobDetail for our annotated job class. As a dependency was added external library which adds actuator endpoint for managing quartz jobs and triggers.

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by Vladyslav-Frolov — APPROVED** [review]

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/config/QuartzBeanPostProcessor.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import org.quartz.Job;
+import org.quartz.JobDetail;
+import org.quartz.Trigger;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.context.annotation.Configuration;
+
+
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/config/QuartzBeanPostProcessor.java]: Please remove extra lines

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/config/QuartzUtil.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import lombok.extern.slf4j.Slf4j;
+import org.quartz.CronTrigger;
+import org.quartz.Job;
+import org.quartz.JobBuilder;
+import org.quartz.JobDetail;
+import org.quartz.TriggerBuilder;
+
+@Slf4j
+public class QuartzUtil {
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/config/QuartzUtil.java]: Please make it final as well

File: home-application/pom.xml [hunk 3]
```diff
@@ -73,6 +73,11 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-actuator</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.quartz-actuator</groupId>
+            <artifactId>quartz-actuator-application</artifactId>
+            <version>1.2.2</version>
```
Comment (kh0ma) [inline home-application/pom.xml]: Please use the property for defining version to be at least consistent with other dependencies.

File: pom.xml [hunk 4]
```diff
@@ -97,6 +98,16 @@
         </repository>
     </distributionManagement>
 
+    <repositories>
+        <repository>
+            <id>github</id>
+            <name>GitHub Illia-Shchur Apache Maven Packages</name>
+            <url>https://Illia-Shchur:${github.quartz-actuator.secret}@maven.pkg.github.com/Illia-Shchur/quartz-actuator</url>
+            <releases><enabled>true</enabled></releases>
```
Comment (kh0ma) [inline pom.xml]: Isn't it enabled by default as well as snapshots?

File: pom.xml [hunk 5]
```diff
@@ -43,6 +43,7 @@
         <checkstyle.configLocation>${maven.multiModuleProjectDirectory}/.github/checkstyle-configuration.xml
         </checkstyle.configLocation>
         <jacoco.version>0.8.6</jacoco.version>
+        <github.quartz-actuator.secret>${Dapi.application.dependency.quartz-actuator}</github.quartz-actuator.secret>
```
Comment (kh0ma) [inline pom.xml]: `<github-packages-read-token></github-packages-read-token>`

Comment (kh0ma) [reply pom.xml]: And additionally:
`<github-packages-read-username></github-packages-read-username>`

File: pom.xml [hunk 6]
```diff
@@ -97,6 +98,16 @@
         </repository>
     </distributionManagement>
 
+    <repositories>
+        <repository>
+            <id>github</id>
+            <name>GitHub Illia-Shchur Apache Maven Packages</name>
+            <url>https://Illia-Shchur:${github.quartz-actuator.secret}@maven.pkg.github.com/Illia-Shchur/quartz-actuator</url>
```
Comment (kh0ma) [inline pom.xml]: `<url>https://${github-packages-read-username}:${github-packages-read-token}@maven.pkg.github.com/orgs/Home-Project-Engineering/*</url>`

File: pom.xml [hunk 7]
```diff
@@ -97,6 +98,16 @@
         </repository>
     </distributionManagement>
 
+    <repositories>
+        <repository>
+            <id>github</id>
+            <name>GitHub Illia-Shchur Apache Maven Packages</name>
```
Comment (kh0ma) [inline pom.xml]: `<name>GitHub Home-Project-Engineering Apache Maven Packages</name>`

File: home-application/pom.xml [hunk 8]
```diff
@@ -73,6 +73,11 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-actuator</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.quartz-actuator</groupId>
```
Comment (kh0ma) [inline home-application/pom.xml]: Let the group id be `tech.home-project-engineering`

File: home-application/pom.xml [hunk 9]
```diff
@@ -73,6 +73,11 @@
...
+            <artifactId>quartz-actuator-application</artifactId>
```
Comment (kh0ma) [inline home-application/pom.xml]: Let's name it `quartz-actuator`

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: .github/workflows/CD.yml [hunk 10]
```diff
@@ -48,6 +48,7 @@ jobs:
           -Ddockerhub.username=${{secrets.DOCKER_USERNAME}}
           -Ddockerhub.password=${{secrets.DOCKER_PASSWORD}}
           -Dapi.tests.verbose.logging=${{env.API_TESTS_VERBOSE_LOGGING}}
+          -Dapi.application.dependency.quartz-actuator=${{secrets.SECRET_QUARTZ_ACTUATOR_LIBRARY}}
```
Comment (kh0ma) [inline .github/workflows/CD.yml]: `-Dgithub.packages.read.username=${{secrets.PACKAGES_READ_USERNAME}}`
`-Dgithub.packages.read.token=${{secrets.PACKAGES_READ_TOKEN}}`

Please also add documentation on how to configure it onto local boxes

**Review by kh0ma — CHANGES_REQUESTED** [review]

**Review by AriannaLiss — APPROVED** [review]

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/invitations/InvitationApiIT.java [hunk 11]
```diff
@@ -8,6 +8,7 @@
 import com.softserveinc.ita.homeproject.api.tests.utils.mail.mock.dto.MailHogApiResponse;
 import com.softserveinc.ita.homeproject.model.*;
 import org.apache.commons.lang3.RandomStringUtils;
+import org.junit.jupiter.api.Disabled;
```
Comment (IhorSamoshost) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/invitations/InvitationApiIT.java:11]: What this import added for?

File: .github/workflows/CI.yml [hunk 12]
```diff
@@ -31,3 +31,5 @@ jobs:
           -Dhome.application.admin.username=${{secrets.HOME_APPLICATION_ADMIN_USERNAME}}
           -Dhome.application.admin.password=${{secrets.HOME_APPLICATION_ADMIN_PASSWORD}}
           -Dapi.tests.verbose.logging=${{env.API_TESTS_VERBOSE_LOGGING}}
+          -Dgithub.packages.read.username=${{secrets.PACKAGES_READ_USERNAME}}
+          -Dgithub.packages.read.token=${{secrets.PACKAGES_READ_TOKEN}}
```
Comment (IhorSamoshost) [inline .github/workflows/CI.yml:35]: Why are here two workflows with the same name? 
it was some kind of glitch. I had it twice)

**Review by IRoss-git — APPROVED** [review]

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/config/QuartzUtil.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+        return TriggerBuilder.newTrigger()
+                .forJob(jobDetail)
+                .withIdentity(triggerName,group)
+                .withSchedule(cronSchedule(cron))
+                .build();
+    }
+
+    public static JobDetail createJobDetail(Class<? extends Job> jobClass, String jobName,String group) {
+        log.debug("createJobDe tail(jobClass={}, jobName={})", jobClass.getName(), jobName);
```
Comment (IhorSamoshost) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/config/QuartzUtil.java:30]: extra space in a word "createJobDe tail"

File: .github/workflows/CI.yml (hunk 12 above)
Comment (Illia-Shchur) [reply .github/workflows/CI.yml:35]: We need those two params a both of our workflows in order to get dependency from GitHub.

Comment (Illia-Shchur) [reply .github/workflows/CI.yml:35]: We need those two params in both of our workflows in order to get dependency from GitHub.

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/config/QuartzJobBeanAutoConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.quartz.Trigger;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+/**
+ * QuartzJobBeanAutoConfiguration annotation for @see org.quartz.Job
+ * For Job annotated with QuartzJobBeanAutoConfiguration will be generated
```
Comment (likeRewca) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/config/QuartzJobBeanAutoConfiguration.java:14]: Maybe will be better if you write javadoc under annotation

**Review by kh0ma — APPROVED** [review]

**Pull request merged**